### ISSUE TITLE
Direct Uploads

### ISSUE THREAD

**Issue by ladar** [hashicorp/vagrant_cloud#72]
@chrisroberts hi Chris, I was having trouble uploading the `3.2.10` release, and stumbled across your new "direct upload" option. So I gave it a try. My initial tests seemed went smoothly, although I did noticed a short lag between the final callback, and when the box could be downloaded.

But with high hopes, I kicked off a bulk upload with multi-uploads running in parallel. On the bright side, the process went relatively fast (only had 1 gig link to test with), but I was able to upload all 622 box files in ~90 minutes, without any errors on the file upload, or the call back. Unfortunately, it didn't seem to work. No matter how long i waited, I couldn't retrieve any of the box files that were uploaded. Eventually I deleted everything, and restarted the upload using the old strategy.... 1 box at a time, with a 30 second delay between each upload. 

Is there something I'm missing? I based my upload script off the ruby code in this repo, and this documentation:

[https://www.vagrantup.com/vagrant-cloud/api#upload-a-provider-directly-to-backend-storage](url)

My current script is at:

[https://github.com/lavabit/robox/blob/ce2ad9f3bff11ac04f9890ffee294ed90e95ee57/res/scripts/direct.sh](url)

Do I need to add a delay between the file upload and the callback? Or some other step?

**Comment by chrisroberts** [comment]
Hey,

I'd need to see some debug output to determine if there was some error being encountered during the callback that was ultimately causing the issue. Does the uploaded asset not show up in the Vagrant Cloud UI after the upload and callback are completed? Or is it just the accessibility when downloading? Compared to the `uploader.sh` script you have, there's no release call happening at the end in this `direct.sh` script which is my best guess as to why you aren't able to download them.

**Comment by ladar** [comment]
@chrisroberts lately I've been using the `silent.sh` script, and then calling the `release.sh` when all the boxes are uploaded. For my attempt yesterday, it was a re-upload, because several of the boxes failed during the initial attempt. So the boxes were already 'released' ... but I ran the release step again, before giving up, just to be sure.

Is there a way for me to enable debug output on the callback PUT request? I tried running it manually, put it didn't appear to provide any output. All I could get via the `curl` debug output, was that the server returned an HTTP 200 code.

As I recall, the web interface showed the box version, and providers, but didn't list the box size, something that seems to happen when an upload fails. 

If you'd like, I can rerun the upload as version `3.2.9` ... I skip over the odd numbered version number version numbers so I can upload experimental boxes for testing, without them getting automatically selected. Let me know?

**Comment by chrisroberts** [comment]
Was able to get a good reproduction of what you were experiencing. Turns out there was a delay occurring depending on the timing due to the asset verification job getting dumped into the low priority job queue. Bumped the job's priority up and that should clear things up. If you want to give your uploads a run again using the direct method, that would be super for an extra confirmation that this is cleared up. Thanks!

**Comment by ladar** [comment]
I'll kick it off now. I'll upload the Lavabit/Magma and Lineage boxes first, one at a time. And if that works, I'll try the Generic/Roboxes in parallel (as v3.2.9).

**Comment by ladar** [comment]
From what I can tell, and based on spot checking some of the uploaded boxes, it worked great! I forgot to benchmark the upload process, but I should have plenty of chances to do that going forward. Based on my tests yesterday though, I think this will cut the time it takes to upload a new Robox release down from its current ~18-24 hours,  to just ~1-2 hours. It will probably be even less once I switch the staging server over to a 10 gig port. 

Thanks @chrisroberts!
